[sammy] eta-expansion, overloading, existentials

adriaanm opened the pull request:
Review by @retronym, please.

While playing with Java 8 Streams from the repl,
I realized we weren't eta-expanding,
nor resolving overloading properly for SAMs.

Also, the way Java uses wildcards to represent use-site variance
stresses type inference (since it results in existentials).

Introduce `wildcardExtrapolation` to simplify the resulting types
(without losing precision): `wildcardExtrapolation(tp) =:= tp`.

For example, the `MethodType` given by `def bla(x: (_ >: String)): (_ <: Int)`
is both a subtype and a supertype of `def bla(x: String): Int`.

Translating http://winterbe.com/posts/2014/07/31/java8-stream-tutorial-examples/ into Scala:

```
scala> import java.util.Arrays
scala> import java.util.stream.Stream
scala> import java.util.stream.IntStream
scala> val myList = Arrays.asList("a1", "a2", "b1", "c2", "c1")
myList: java.util.List[String] = [a1, a2, b1, c2, c1]

scala> myList.stream.filter(_.startsWith("c")).map(_.toUpperCase).sorted.forEach(println)
C1
C2

scala> myList.stream.filter(_.startsWith("c")).map(_.toUpperCase).sorted
res8: java.util.stream.Stream[?0] = java.util.stream.SortedOps$OfRef@133e7789

scala> Arrays.asList("a1", "a2", "a3").stream.findFirst.ifPresent(println)
a1

scala> Stream.of("a1", "a2", "a3").findFirst.ifPresent(println)
a1

scala> IntStream.range(1, 4).forEach(println)
<console>:37: error: object creation impossible, since method accept in trait IntConsumer of type (x$1: Int)Unit is not defined
(Note that Int does not match Any: class Int in package scala is a subclass of class Any in package scala, but method parameter types must match exactly.)
              IntStream.range(1, 4).forEach(println)
                                            ^

scala> IntStream.range(1, 4).forEach(println(_: Int)) // TODO: can we avoid this annotation?
1
2
3

scala> Arrays.stream(Array(1, 2, 3)).map(n => 2 * n + 1).average.ifPresent(println(_: Double))
5.0

scala> Stream.of("a1", "a2", "a3").map(_.substring(1)).mapToInt(_.parseInt).max.ifPresent(println(_: Int)) // whoops!
ReplGlobal.abort: Unknown type: <error>, <error> [class scala.reflect.internal.Types$ErrorType$, class scala.reflect.internal.Types$ErrorType$] TypeRef? false
error: Unknown type: <error>, <error> [class scala.reflect.internal.Types$ErrorType$, class scala.reflect.internal.Types$ErrorType$] TypeRef? false
scala.reflect.internal.FatalError: Unknown type: <error>, <error> [class scala.reflect.internal.Types$ErrorType$, class scala.reflect.internal.Types$ErrorType$] TypeRef? false
    at scala.reflect.internal.Reporting$class.abort(Reporting.scala:59)

scala> IntStream.range(1, 4).mapToObj(i => "a" + i).forEach(println)
a1
a2
a3

```

adriaanm committed:
[sammy] eta-expansion, overloading (SI-8310)

Playing with Java 8 Streams from the repl showed
we weren't eta-expanding, nor resolving overloading for SAMs.

Also, the way Java uses wildcards to represent use-site variance
stresses type inference past its bendiness point (due to excessive existentials).

I introduce `wildcardExtrapolation` to simplify the resulting types
(without losing precision): `wildcardExtrapolation(tp) =:= tp`.

For example, the `MethodType` given by `def bla(x: (_ >: String)): (_ <: Int)`
is both a subtype and a supertype of `def bla(x: String): Int`.

Translating http://winterbe.com/posts/2014/07/31/java8-stream-tutorial-examples/
into Scala shows most of this works, though we have some more work to do (see near the end).

```
scala> import java.util.Arrays
scala> import java.util.stream.Stream
scala> import java.util.stream.IntStream
scala> val myList = Arrays.asList("a1", "a2", "b1", "c2", "c1")
myList: java.util.List[String] = [a1, a2, b1, c2, c1]

scala> myList.stream.filter(_.startsWith("c")).map(_.toUpperCase).sorted.forEach(println)
C1
C2

scala> myList.stream.filter(_.startsWith("c")).map(_.toUpperCase).sorted
res8: java.util.stream.Stream[?0] = java.util.stream.SortedOps$OfRef@133e7789

scala> Arrays.asList("a1", "a2", "a3").stream.findFirst.ifPresent(println)
a1

scala> Stream.of("a1", "a2", "a3").findFirst.ifPresent(println)
a1

scala> IntStream.range(1, 4).forEach(println)
<console>:37: error: object creation impossible, since method accept in trait IntConsumer of type (x$1: Int)Unit is not defined
(Note that Int does not match Any: class Int in package scala is a subclass of class Any in package scala, but method parameter types must match exactly.)
              IntStream.range(1, 4).forEach(println)
                                            ^

scala> IntStream.range(1, 4).forEach(println(_: Int)) // TODO: can we avoid this annotation?
1
2
3

scala> Arrays.stream(Array(1, 2, 3)).map(n => 2 * n + 1).average.ifPresent(println(_: Double))
5.0

scala> Stream.of("a1", "a2", "a3").map(_.substring(1)).mapToInt(_.parseInt).max.ifPresent(println(_: Int)) // whoops!
ReplGlobal.abort: Unknown type: <error>, <error> [class scala.reflect.internal.Types$ErrorType$, class scala.reflect.internal.Types$ErrorType$] TypeRef? false
error: Unknown type: <error>, <error> [class scala.reflect.internal.Types$ErrorType$, class scala.reflect.internal.Types$ErrorType$] TypeRef? false
scala.reflect.internal.FatalError: Unknown type: <error>, <error> [class scala.reflect.internal.Types$ErrorType$, class scala.reflect.internal.Types$ErrorType$] TypeRef? false
	at scala.reflect.internal.Reporting$class.abort(Reporting.scala:59)

scala> IntStream.range(1, 4).mapToObj(i => "a" + i).forEach(println)
a1
a2
a3

```

adriaanm commented:
/cc FYI, @ichoran

adriaanm commented:
I'll investigate the failures in the REPL transcript later. Need to get back to my talk :)

gourlaysama commented:
:+1: Yay! Overloading is _the_ thing that prevented SAMs from being really useful when calling into java.

BTW there was a ticket for that: [SI-8310](https://issues.scala-lang.org/browse/SI-8310)

adriaanm commented:
Ugh, the jenkins node I added is running an outdated version of java.

adriaanm commented:
Thanks for  the pointer @gourlaysama! I'll include the test case -- it passes now.

retronym commented:
I offlined them about 15 mins ago.

adriaanm commented:
Thanks -- I push --forced to address review feedback (include SI-8310 test case).

adriaanm committed:
[sammy] use correct type for method to override

Don't naively derive types for the single method's signature
from the provided function's type, as it may be a subtype
of the method's MethodType.

Instead, once the sam class type is fully defined, determine
the sam's info as seen from the class's type, and use those
to generate the correct override.

```
scala> Arrays.stream(Array(1, 2, 3)).map(n => 2 * n + 1).average.ifPresent(println)
5.0

scala> IntStream.range(1, 4).forEach(println)
1
2
3
```

Also, minimal error reporting

Can't figure out how to do it properly, but some reporting
is better than crashing. Right? Test case that illustrates
necessity of the clumsy stop gap `if (block exists (_.isErroneous))`
enclosed as `sammy_error_exist_no_crash`

added TODO for repeated and by-name params

adriaanm commented:
Need to update that check file. The new error message is correct, though suboptimal...

retronym commented on src/compiler/scala/tools/nsc/typechecker/Infer.scala:
```diff
@@ -295,11 +295,17 @@ trait Infer extends Checkable {
       (    (tp1 weak_<:< pt)
         || isCoercible(tp1, pt)
         || isCompatibleByName(tp, pt)
+        || isCompatibleSam(tp, pt)
```
I don't see where we predicate this on `-Xexperimental`. I believe we should. I'd also be interested in performance impact of this; we should have a fast path for non-SAMs.

retronym replied:
Oh wait, I see `samOf` checks the settings. That will do.

adriaanm replied:
I figured it's easier to flip the switch if it's only tested in one central spot.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -2818,6 +2849,11 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
 
+      // TODO: improve error reporting -- when we're in silent mode (from `silent(_.doTypedApply(tree, fun, args, mode, pt)) orElse onError`)
+      // the errors in the function don't get out...
+      if (block exists (_.isErroneous))
```
Traversing the typechecked expansion in search of errors seems a bit wasteful. If this is erroneous, there will have been an error issues (admittedly about synthetic code). Would be better to have this explanation automatically attached as an addendum to any errors issued during typechecking the expansion (in a manner similar to names/defaults.)

adriaanm replied:
The problem is that we don't get any errors and the back-end freaks out. I have spent some time debugging this, but couldn't figure it out. I agree this isn't ready for production, but I think it's better than a crash for now.

retronym replied:
Fair enough. Can you include a test case that gets to this spot so we can easily pick up the investigation later?

adriaanm replied:
Will do.

retronym commented on src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:
```diff
@@ -422,6 +422,22 @@ private[internal] trait TypeMaps {
+    def apply(tp: Type): Type =
+      tp match {
+        case BoundedWildcardType(TypeBounds(lo, AnyTpe))  if variance.isContravariant =>lo
+        case BoundedWildcardType(TypeBounds(NothingTpe, hi)) if variance.isCovariant => hi
```
whitespace could be prettified here.

retronym replied:
I wonder if a more general version would also be correct?

``` scala
  case BoundedWildcardType(TypeBounds(lo, _)) if variance.isContravariant => lo
  case BoundedWildcardType(TypeBounds(_, hi)) if variance.isCovariant     => hi
```

adriaanm replied:
No, that wouldn't preserve type equivalence. The reason `def bla(x: (_ >: A)): (_ <: B)` and `def bla(x: A): B` are equivalent is that the existentials are redundant: they encode the variance position of the type of the argument and the result. 

Consider the corresponding function type as a whole: when you request a value of `String => Int`, you're happy to get any `T => Int`, as long as it's for some `T >: String` (because you intend to pass it a `String`, but it will accept any super type of `String`). When you request a `_ >: String => Int`, you're happy to receive a `String => Int` (because your existential conundrum has been resolved). Now, replace 'requesting' by 'assigning type' and 'happiness' by 'subtyping', and I rest my case.

More formally, I have to appeal to the interpretation of variant type parameters in terms of bounded abstract type members (see one-but-last page of http://lampwww.epfl.ch/~odersky/papers/ScalableComponent.pdf):

```
class C
trait X {type T}

// contra: trait X[-A] == X{ type T >: A}
typeOf[X{ type T >: C }] <:< typeOf[X{ type T >: C }] // true      typeOf[X[C]] <:< typeOf[X[_ >: C]]
typeOf[X{ type T >: C }] <:< typeOf[X{ type T >: C }] // true      typeOf[X[_ >: C]] <:< typeOf[X[C]]

// cov: trait X[+A] == X{ type T <: A}
typeOf[X{ type T <: C }] <:< typeOf[X{ type T >: C }] // false     typeOf[X[C]] <:< typeOf[X[_ >: C]]
typeOf[X{ type T >: C }] <:< typeOf[X{ type T <: C }] // false     typeOf[X[_ >: C]] <:< typeOf[X[C]]

// inv: trait X[A] == X{ type T = A }, or X{ type T = >: A <: A }
typeOf[X{ type T >: C <: C }] <:< typeOf[X{ type T >: C }] // true  typeOf[X[C]] <:< typeOf[X[_ >: C]]
typeOf[X{ type T >: C }] <:< typeOf[X{ type T >: C <: C }] // false typeOf[X[_ >: C]] <:< typeOf[X[C]]
```

retronym commented on src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:
```diff
@@ -422,6 +422,22 @@ private[internal] trait TypeMaps {
 
+  /**
+   * Get rid of BoundedWildcardType where variance allows us to do so.
+   * Invariant: `wildcardExtrapolation(tp) =:= tp`
```
Some unit tests to show this would be handy.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -741,6 +741,26 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+      }
+
+      if (samSym.exists && samSym.owner != correspondingFunctionSymbol) // don't treat Functions as SAMs
+        wildcardExtrapolation(normalize(tp memberInfo samSym))
```
Why normalize?

adriaanm replied:
to turn it into a function type

retronym commented on test/files/pos/sammy_exist.scala:
```diff
@@ -0,0 +1,24 @@
+//
+// def jf[T, R](f: T => R): SFun[T, R] = (x: T) => f(x): R
+//
+// val aBlaSAM = (new S(new Bla)).map(jf(_.foo)) // : Bla should be inferred [type checks, but existential inferred]
```
These notes aren't needed any more, right?

adriaanm replied:
Nope -- I'll prune.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -2848,15 +2884,11 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
        * Note that the arity of the sam must correspond to the arity of the function.
        */
       val samViable = sam.exists && sameLength(sam.info.params, fun.vparams)
+      val ptNorm = if (samViable) samToFunctionType(pt, sam) else pt
```
This will limit us to SAMs of 22 parameters, if we're happy with that restriction we'll at least need a good error message / test.

retronym replied:
Which reminds me, we also need to consider by-name and repeated parameters, which are under-specced when it comes to anonymous function params, and have been a source of bugs.

With 2.11.4:

```
scala> trait LazySink { def accept(a: => Any): Unit }
defined trait LazySink

scala> val f: LazySink = (a) => (a, a)
f: LazySink = $anonfun$1@1fb26910

scala> f(println("!"))
<console>:10: error: LazySink does not take parameters
              f(println("!"))
               ^

scala> f.accept(println("!"))
!
!
```

This looks like a bug:

```
scala> trait RepeatedSink { def accept(a: Any*): Unit }
defined trait RepeatedSink

scala> val f: RepeatedSink = (a) => println(a)
f: RepeatedSink = $anonfun$1@4799abc2

scala> f.accept(1)
WrappedArray(WrappedArray(1))

scala> f.accept(1, 2)
WrappedArray(WrappedArray(1, 2))
```

retronym replied:
This bug comes from:

``` scala
Apply(Ident(bodyName), fun.vparams map (p => Ident(p.name)))
```

Rather than something like:

``` scala
// TreeGen
  def mkForwarder(target: Tree, vparamss: List[List[Symbol]]) =
    (target /: vparamss)((fn, vparams) => Apply(fn, vparams map paramToArg))
```

adriaanm replied:
I'll have to defer this one. I tried `Apply(Ident(bodyName), fun.vparams map gen.paramToArg)`, but no dice. I'm not sure how we would handle CBN params, but I guess we should support repeated ones somehow if java does?

retronym replied:
Fair enough. I tried this and it seems to work and pass `partest-ack sam`, but we can come back to it.

```
% git diff head
diff --git a/src/compiler/scala/tools/nsc/typechecker/Typers.scala b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
index 5eee600..48345ab 100644
--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -2848,7 +2848,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           Nil,
           List(fun.vparams),
           TypeTree(samMethTp.finalResultType) setPos sampos.focus,
-          Apply(Ident(bodyName), fun.vparams map (p => Ident(p.name)))
+          Apply(Ident(bodyName), fun.vparams map gen.paramToArg)
         )

       val serializableParentAddendum =

% cat sandbox/test.scala
trait RepeatedSink { def accept(a: Any*): Unit }

object Test {
  def main(args: Array[String]): Unit = {
    val f: RepeatedSink = (a) => println(a)
    f.accept(1)
  }
}
% qscalac -Xexperimental sandbox/test.scala && qscala Test
WrappedArray(1)
```

adriaanm replied:
Sorry, not sure how I botched trying that. You're right. Added a commit.

adriaanm commented:
BTW, https://gist.github.com/adriaanm/892d6063dd485d7dd221

adriaanm committed:
[sammy] support repeated params

Generate correct trees to refer to repeated params using `gen.paramToArg`.
Based on retronym's review feedback.

retronym commented:
LGTM

VladUreche commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -2701,6 +2725,22 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
      * which must be fully defined. Would be nice to have some kind of mechanism to insert type vars in a block of code,
      * and have the instantiation of the first occurrence propagate to the rest of the block.
+     *
+     * TODO: by-name params
```
@adriaanm, by this you mean automatically applying the SAM implementation's abstract method?

adriaanm replied:
Not sure what you're referring to, but I meant that `f(println("!"))` should behave the same as `f.accept(println("!"))` below.

VladUreche replied:
Got it now, thanks @adriaanm!

VladUreche commented:
Cool patch! I'm super excited about this making its way to 2.11.5 :+1:
